File: development/matlab/nlvd_dataset.m
classdef nlvd_dataset < handle

    properties (GetAccess=public, SetAccess=protected)
        dataset_name
        subset_name
        split_ids
        im2rid
        rid2r
        tid2p
        level1
        level2
        toolbox_path
    end

    methods (Access=public)

        function obj = nlvd_dataset(dataset_name, subset_name)
            obj.toolbox_path = fileparts(fileparts(fileparts(mfilename('fullpath'))));
            if ~exist('dataset_name', 'var') || isempty(dataset_name)
                dataset_name = 'vg_v1';
            end
            if ~exist('subset_name', 'var') || isempty(subset_name)
                subset_name = 'train';
            end
            obj.dataset_name = dataset_name;
            obj.subset_name = subset_name;
            data_dir = fullfile( ...
                    obj.toolbox_path, 'annotations', obj.dataset_name);
            fprintf('Annotatoin directory: %s\n', data_dir);

            fprintf('Loading region descriptions file ... \n');
            if ~strcmp(obj.subset_name, 'train')
                region_mat_file = sprintf('%s/%s_region_description.mat', data_dir, subset_name);
                tmp = load(region_mat_file);
                region_description = tmp.region_description;
                obj.im2rid = region_description.im2rid;
                obj.tid2p = region_description.tid2p;
                obj.rid2r = region_description.rid2r;
            else
                [obj.im2rid, obj.rid2r, obj.tid2p] = load_train_region_description(data_dir);
            end

            fprintf('Loading subset split file ... \n');
            split_file = sprintf('%s/subset_splits.json',data_dir);
            fid = fopen(split_file);
            raw = fread(fid,inf);
            str = char(raw');
            fclose(fid);
            obj.split_ids = jsondecode(str);
                       
            if strcmp(obj.subset_name, 'test')
                fprintf('Loading level1 query file ... \n');
                level1_file = sprintf('%s/level1_im2p.json', data_dir);
                fid = fopen(level1_file);
                raw = fread(fid,inf);
                str = char(raw');
                fclose(fid);
                obj.level1 = jsondecode(str);

                fprintf('Loading level2 query file ...\n');
                level2_file = sprintf('%s/level2_im2p.json', data_dir);
                fid = fopen(level2_file);
                raw = fread(fid,inf);
                str = char(raw');
                fclose(fid);
                obj.level2 = jsondecode(str);
            end
        end
        
        function phrases = text_id_to_phrase(obj, text_ids)
            phrases = [];
            for k =1:numel(text_ids)
                fieldname = sprintf('x%d', text_ids(k));
                phrases= [phrases; {obj.tid2p.(fieldname)}];
            end
        end
        
        function image_ids = image_ids_in_subset(obj)
            image_ids = obj.split_ids.(obj.subset_name);
        end
        
        function image_info = annotation(obj, image_id)
            image_info.image_id = image_id;
            image_dir = fullfile( ...
                                 obj.toolbox_path, 'images', obj.dataset_name);
            if(exist(sprintf('%s/%d.jpg', image_dir, image_id), 'file'))
                image_info.image_path = sprintf('%s/%d.jpg', image_dir, image_id);
            else
                image_info.image_path = sprintf('%d.jpg', image_id);
            end
            image_info.regions = [];
            fieldname = sprintf('x%d', image_id);
            if(isfield(obj.im2rid, fieldname))
                region_ids = obj.im2rid.(fieldname);
            else
                fprintf('no region information exist\n');
            end
            for k = 1:numel(region_ids)
                rid = region_ids(k);
                fieldname = sprintf('x%d', rid);
                region_info = obj.rid2r.(fieldname);
                %here the coordinate of bounding box is 1-based
                region_info.x = region_info.x+1;
                region_info.y = region_info.y+1;
                region_info.phrase = obj.tid2p.(sprintf('x%d', region_info.phrase_id));
                image_info.regions = [image_info.regions; region_info];
            end
        end
        
        function text_ids = test_text_ids(obj, image_id, level_id)
            % every subset should support 
            if ~exist('level_id', 'var') || isempty(level_id)
                level_id = 0;
            end
            if level_id > 0
                assert(strcmp(obj.subset_name, 'test'), ...
                    'only test set has different difficulty level for detection task\n')
            end
            text_ids = [];
            if(level_id==0)
                fieldname = sprintf('x%d', image_id);
                if(isfield(obj.im2rid, fieldname))
                    region_ids = obj.im2rid.(fieldname);
                else
                    fprintf('no text information exist\n');
                end
                for k = 1:numel(region_ids)
                    rid = region_ids(k);
                    fieldname = sprintf('x%d', rid);
                    region_info = obj.rid2r.(fieldname);
                    test_ids =[test_ids; region_info.phrase_id];
                end
            end
            if(level_id==1)
                fieldname = sprintf('x%d', image_id);
                if(isfield(obj.level1, fieldname))
                    text_ids = obj.level1.(fieldname);
                else
                    fprintf('no text information exist\n');
                end
            end
            if(level_id==2)
                fieldname = sprintf('x%d', image_id);
                if(isfield(obj.level1, fieldname))
                    text_ids = obj.level1.(fieldname);
                else
                    fprintf('no text information exist\n');
                end
            end
        end
        
        function test_obj = create_test(obj, test_title, level_id)
            test_dir = fullfile( ...
                obj.toolbox_path, ...
                'results', obj.dataset_name, test_title ...
            );
            test_obj = nlvd_test(obj, test_dir, level_id);
        end
        
    end
end

File: development/python/nlvd_dataset.py
import json
import os.path as osp
from nlvd_test import nlvd_test


class NLVDDataset:

    def __init__(self, dataset_name='vg_v1', subset_name='train'):
        self.dataset_name = dataset_name
        self.root_dir = osp.dirname(osp.dirname(osp.dirname(osp.abspath(__file__))))
        self.data_dir = osp.join(self.root_dir, 'annotations', self.dataset_name)
        self.image_dir = osp.join(self.root_dir, 'images') 
        self.subset_name = subset_name
        print('Loading region description file ...')
        self.raw_data = json.load(open(osp.join(self.data_dir, '%s_region_description.json' % self.subset_name)))
        self.im2rid = self.raw_data['im2rid']
        self.rid2r = self.raw_data['rid2r']
        self.tid2p = self.raw_data['tid2p']
        print('Loading subset split file ...')
        self.raw_split = json.load(open(osp.join(self.data_dir, 'subset_splits.json'), 'r'))
        self.image_ids = self.raw_split.get(subset_name)
        if self.image_ids is None:
            raise ValueError('Please enter valid subset name <train|val|test>')
        if self.subset_name == "test":
            print('Loading level1 query file ...')
            self.level1 = json.load(open(osp.join(self.data_dir, 'level1_im2p.json'), 'r'))
            print('Loading level2 query file ...')
            self.level2 = json.load(open(osp.join(self.data_dir, 'level2_im2p.json'), 'r'))
    
    # return the image ids in the split
    def image_ids_in_subset(self):
        return self.image_ids

    # return a data dict for the query image id
    def annotation(self, image_id):
        data = {'image_id': image_id, 'regions': [],
                'image_path': osp.join(self.image_dir, '%s.jpg' % str(image_id))} 
        rids = self.im2rid[str(image_id)]
        for rid in rids:
            region = self.rid2r[str(rid)]
            region['phrase'] = self.tid2p[str(region['phrase_id'])]
            data['regions'].append(region)
        return data
    
    # return a list of phrases by given phrase ids
    def text_id_to_phrase(self, text_ids):
        phrases = []
        for tid in text_ids:
            phrases.append(self.tid2p[str(tid)])
        return phrases
    
    # return a list of required test phrases given test level and image ids
    def test_text_ids(self, image_id, level_id):
        if level_id == 0:
            text_ids = [self.rid2r[str(rid)]['phrase_id'] for rid in self.im2rid[str(image_id)]]
        elif level_id == 1:
            text_ids = self.level1[str(image_id)]
        elif level_id == 2:
            text_ids = self.level2[str(image_id)]
        else:
            raise ValueError('Please enter valid level <0|1|2>')
        return text_ids 
    
    def create_test(self, test_title, level_id):
        test_dir = osp.join(self.root_dir, 'results', self.dataset_name, test_title)
        return nlvd_test(self, test_dir, level_id)


nlvd_dataset = NLVDDataset


